controllers/kafkamanagement/usercertificate_controller.go

```diff
@@ -0,0 +1,380 @@
+	l := log.FromContext(ctx)
+
+	cert := &kafkamanagementv1beta1.UserCertificate{}
+	err := r.Get(ctx, req.NamespacedName, cert)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if cert.DeletionTimestamp != nil {
+		return ctrl.Result{}, r.handleDelete(ctx, cert)
+	}
+
+	if cert.Status.CertID != "" {
+		return ctrl.Result{}, nil
+	}
+
+	kafkaUser := &kafkamanagementv1beta1.KafkaUser{}
+	err = r.Get(ctx, cert.Spec.UserRef.AsNamespacedName(), kafkaUser)
+	if err != nil {
+		r.EventRecorder.Eventf(cert, models.Warning, models.CreationFailed,
+			"Failed to get KafkaUser resource. Reason: %v", err,
+		)
+		return ctrl.Result{}, err
+	}
+
+	kafkaCluster := &clustersv1beta1.Kafka{}
+	err = r.Get(ctx, cert.Spec.ClusterRef.AsNamespacedName(), kafkaCluster)
+	if err != nil {
+		r.EventRecorder.Eventf(cert, models.Warning, models.CreationFailed,
+			"Failed to get Kafka resource. Reason: %v", err,
+		)
+		return ctrl.Result{}, err
+	}
+
+	res, err := r.checkKafkaUserExistsOnCluster(kafkaCluster, kafkaUser)
+	if err != nil {
+		return res, err
+	}
+
+	if cert.Spec.GenerateCSR != nil && cert.Spec.SecretRef == nil {
+		err = r.createCSR(ctx, cert, kafkaUser)
+		if err != nil {
+			r.EventRecorder.Eventf(cert, models.Warning, models.CreationFailed,
+				"Certificate signing request creation failed. Reason: %v", err,
+			)
+			return ctrl.Result{}, err
+		}
+	}
+
+	err = r.handleCreate(ctx, cert, kafkaCluster, kafkaUser)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	l.Info("Certificate has been created",
+		"certId", cert.Status.CertID,
+	)
+	r.EventRecorder.Event(cert, models.Normal, models.Created, "Certificate has been created")
+
+	return ctrl.Result{}, nil
+}
+
+func (r *UserCertificateReconciler) checkKafkaUserExistsOnCluster(cluster *clustersv1beta1.Kafka, user *kafkamanagementv1beta1.KafkaUser) (ctrl.Result, error) {
+	var userRefExists bool
+	for _, userRef := range cluster.Spec.UserRefs {
+		if userRef.Name == user.Name && userRef.Namespace == user.Namespace {
+			userRefExists = true
+			break
+		}
+	}
+
+	event, clusterEventExists := user.Status.ClustersEvents[cluster.Status.ID]
+
+	if !userRefExists || !clusterEventExists {
+		return ctrl.Result{}, fmt.Errorf("user %s/%s is not added to the cluster %s/%s",
+			user.Namespace, user.Name, cluster.Namespace, cluster.Name)
+	}
+
+	if event == models.CreatingEvent {
+		// Waiting until user is created
+		return ctrl.Result{RequeueAfter: time.Second}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *UserCertificateReconciler) handleCreate(
+	ctx context.Context,
+	cert *kafkamanagementv1beta1.UserCertificate,
+	cluster *clustersv1beta1.Kafka,
+	user *kafkamanagementv1beta1.KafkaUser,
+) error {
+	l := log.FromContext(ctx)
+
+	l.V(1).Info("UserCertificateReconciler::handleCreate")
+
+	userSecret := &v1.Secret{}
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: user.Spec.SecretRef.Namespace,
+		Name:      user.Spec.SecretRef.Name,
```

Comment:
Let's also use `AsNamespacedName()` method here

Reply:
Here we cannot use it because `user` has own reference type `secretRef`